### PR TITLE
BOM-1078

### PR DESCRIPTION
**Pull request by awais786**
- Execute make upgrade
- Fix isort issues
- Migration added for mocked model it was the root cause for sqlite3 failures.
- py3.5 support
- django22 support

BOM-1078

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -2,14 +2,28 @@ language: python
...
 - 2.7
-- 3.6
+- 3.5
 
 services:
 - xvfb
 
 env:
 - DJANGO_SETTINGS_MODULE=test_setting  TOXENV=django111-drf37
+- DJANGO_SETTINGS_MODULE=test_setting  TOXENV=django111-drf38
```
Comment (Ayub-Khan) [inline .travis.yml]: move 'DJANGO_SETTINGS_MODULE=test_setting' to pytest configs from all those envs

File: .travis.yml [hunk 2]
```diff
@@ -20,6 +34,35 @@ env:
 - DJANGO_SETTINGS_MODULE=test_setting  TOXENV=pii_check
 - TOXENV=translations
 
+
+matrix:
+  exclude:
+    - python: 2.7
+      env: DJANGO_SETTINGS_MODULE=test_setting  TOXENV=django20-drf37
```
Comment (Ayub-Khan) [inline .travis.yml]: after moving remove from here.

File: .travis.yml [hunk 3]
```diff
@@ -52,4 +95,4 @@ deploy:
     secure: Bs6z6iFpO86mav8qo3cDhcjDckNXDGfcbuZtSWbbrIFe9Q0VSVz046s9Yo7hHjjaRUSb6frlTC8lYRNDZ2m7TxXrh4vXHrHsr9ajcJXUZJCwqnnxIEFeHe02hD/7RPJ3xK7tkGCxd/2AUHEHXYylGOePDc/Ck/2VYmGQ0EijcfZeOJ5macZ65TP0Kwvf+IDOko++W94f3Ujglu+lWryBTR61Yaf7zt7fKQVgjxGQjb07F1ze39A00xkYYYbIDSrzjZeebs/pzCjZhrrkzTwReDItTxk9K7cEx5nlRvQul6Wj5USe1uBL5SodP2XCp3mi7XrB/LOt6kDEcdREUzqZUqB5n21VVey0JWihd579PIHFhTXDQpdSJNhOi86Mv8lj3v/XUTRaZBV+2OPVGKSRgmJ0Qv7BkQXOp8tzSN4AINrYD2bMod3jVaGcjg7MYUpJ7B2swWUSa0aLLUe1M2RYzfQ2CiGeN9jCdz8jmQie+HGc5CkaMFtoT1OUt1R7wuO2RNo8pYrP0KJTWpOCHlGn+5r17iDOPXoc/m6a5wN3sxMrARqF5A/PRv3uY/Nb+4bfTeczDZudpxVxGFxe85rahXjh7tSHX5JOiaOXDO9xWj2EBRKKeD4xbgRSBZvjnRrZ0f9ygZjuzFv86KixwYI3HFo21SJQmIaJIUSOoM+r1lQ=
   on:
     tags: true
-    condition: '$TOXENV=quality'
+    condition: '$TOXENV=quality'
```
Comment (Ayub-Khan) [inline .travis.yml]: new line.

File: edx_proctoring/backends/tests/test_software_secure.py [hunk 4]
```diff
@@ -6,15 +6,13 @@
...
 import ddt
-from httmock import HTTMock, all_requests
-from mock import MagicMock, patch
-
+import simplejson as json
 from django.contrib.auth.models import User
 from django.test import TestCase
 from django.urls import reverse
+from httmock import HTTMock, all_requests
+from mock import MagicMock, patch
```
Comment (Ayub-Khan) [inline edx_proctoring/backends/tests/test_software_secure.py]: mock is core library should should not be with third party.

File: edx_proctoring/management/commands/tests/test_set_attempt_status.py [hunk 5]
```diff
@@ -7,9 +7,8 @@
 from datetime import datetime
 
 import pytz
-from mock import MagicMock, patch
-
 from django.core.management import call_command
+from mock import MagicMock, patch
```
Comment (Ayub-Khan) [inline edx_proctoring/management/commands/tests/test_set_attempt_status.py]: same as above. mock is core library should not be with third party.

File: edx_proctoring/tests/test_api.py [hunk 6]
```diff
@@ -12,43 +12,86 @@
 
 import ddt
 import pytz
-
-from freezegun import freeze_time
-from mock import MagicMock, patch
 import six
 from six.moves import range
+from freezegun import freeze_time
+from mock import MagicMock, patch
```
Comment (Ayub-Khan) [inline edx_proctoring/tests/test_api.py]: mock is core library should should not be with third party.

File: edx_proctoring/tests/test_workerconfig.py [hunk 7]
```diff
@@ -51,6 +50,7 @@ def _check_outfile(self, expected):
         if expected is None:
             self.assertFalse(os.path.exists(self.outfile))
         else:
+
```
Comment (Ayub-Khan) [inline edx_proctoring/tests/test_workerconfig.py]: remove

File: requirements/base.in [hunk 8]
```diff
@@ -21,3 +20,4 @@ edx-opaque-keys>=0.4
 edx-drf-extensions
 edx-rest-api-client>=1.9.2
 edx-when>=0.1.3
+simplejson
```
Comment (Ayub-Khan) [inline requirements/base.in]: new line.

File: setup.cfg [hunk 9]
```diff
@@ -1,10 +1,15 @@
...
+multi_line_output=3
 known_edx =
-known_django = django
 known_djangoapp = model_utils
 known_first_party = edx_proctoring
 sections = FUTURE,STDLIB,THIRDPARTY,DJANGO,DJANGOAPP,EDX,FIRSTPARTY,LOCALFOLDER
+known_third_party=django,rest_framework,ddt,mock,waffle,six,crum,httmock,edx_rest_framework_extensions,pytz,edx_when,eventtracking,opaque_keys,edx_rest_api_client,webpack_loader,jwt,freezegun
+skip=
+    envs
+    migrations
```
Comment (Ayub-Khan) [inline setup.cfg]: skind all '.*'

File: setup.py [hunk 10]
```diff
@@ -72,13 +72,13 @@ def is_requirement(line):
         'Development Status :: 4 - Beta',
         'Framework :: Django',
         'Framework :: Django :: 1.11',
```
Comment (Ayub-Khan) [inline setup.py]: django 20, 21,

File: tox.ini [hunk 11]
```diff
@@ -1,7 +1,6 @@
 [tox]
 envlist =
-        py{27,36}-django{111}-drf{37,38,39},
-        py36-django111-drflatest,
+        py{27,35}-django{111}-drf{37,38,39,latest},py{35,36}-django{20,21,22}-drf{37,38,39,latest},
```
Comment (Ayub-Khan) [inline tox.ini]: remove 36 for the time being already too many tests

File: tox.ini [hunk 12]
```diff
@@ -10,13 +9,16 @@ envlist =
...
+    django21: Django>=2.1,<2.2
+    django22: Django>=2.2,<2.3
     drf37: djangorestframework<3.8.0
     drf38: djangorestframework<3.9.0
     drf39: djangorestframework<3.10.0
     drflatest: djangorestframework
     -rrequirements/test.txt
 commands =
-    py.test -rfe --cov=edx_proctoring --cov-report=html --ds=test_settings -n 3 {posargs}
+    python -Wd -m pytest -rfe --cov=edx_proctoring --cov-report=html --ds=test_settings -n 3 {posargs}
```
Comment (Ayub-Khan) [inline tox.ini]: use pytest config block for configuration and django settings

File: mock_apps/migrations/0001_initial.py [hunk 13]
```diff
@@ -0,0 +1,25 @@
+# Generated by Django 2.2.8 on 2019-12-12 08:19
```
Comment (awais786) [inline mock_apps/migrations/0001_initial.py]: in django22 tests were failing on sqlite3.

File: edx_proctoring/apps.py [hunk 14]
```diff
@@ -5,11 +5,11 @@
 
 from __future__ import absolute_import
 
-import json
 import os
 import os.path
 import warnings
 
+import simplejson as json
```
Comment (jmbowman) [inline edx_proctoring/apps.py]: Did you hit a problem that required adding this, or is it just to reduce the odds of encountering a bytes vs. str issue in production?  We're currently on track to have edx-platform on Python 3 in production tomorrow, so not sure it makes sense to make more proactive changes of that sort.

File: edx_proctoring/backends/tests/test_rest.py [hunk 15]
```diff
@@ -3,19 +3,23 @@
 """
 from __future__ import absolute_import
 
-import json
+import responses
+from mock import patch
```
Comment (jmbowman) [inline edx_proctoring/backends/tests/test_rest.py]: This ordering seems odd, not sure why isort moved these.  The `mock` one I can understand if it was run under Python 3, but `responses` should be a 3rd-party package under all Python versions.

File: edx_proctoring/models.py [hunk 16]
```diff
@@ -196,6 +200,14 @@ def delete(self, *args, **kwargs):  # pylint: disable=arguments-differ
         """
         raise NotImplementedError()
 
+    def __str__(self):
+        """ String representation """
```
Comment (jmbowman) [inline edx_proctoring/models.py]: Was there a warning about these being missing?  Doesn't seem like it should have been needed for Django 2.2 compatibility.

File: edx_proctoring/tests/test_services.py [hunk 17]
```diff
@@ -6,12 +6,12 @@
 
 from __future__ import absolute_import
 
+import six
 import types
 import unittest
 from datetime import datetime, timedelta
 
 import pytz
-import six
```
Comment (jmbowman) [inline edx_proctoring/tests/test_services.py]: This move definitely seems wrong, six isn't a standard library package.

File: setup.cfg [hunk 18]
```diff
@@ -1,11 +1,19 @@
 [isort]
-line_length = 120
+indent='    '
+line_length=120
+multi_line_output=3
+known_standard_library=mock,six,time,responses
```
Comment (jmbowman) [inline setup.cfg]: Here's the source of the odd sorting; of these, only `time` is actually a standard library package under Python 2 (and two of them aren't under Python 3 either).

File: requirements/base.in [hunk 19]
```diff
@@ -2,8 +2,7 @@
 -c constraints.txt
 
 # Django/Framework Packages
-django<2,>=1.11
-django-model-utils>=2.3.1
```
Comment (jmbowman) [inline requirements/base.in]: Why was this removed?  It's a direct dependency, used in `edx_proctoring/models.py`.

File: requirements/dev.txt [hunk 20]
```diff
@@ -7,7 +7,6 @@
 argparse==1.4.0           # via caniusepython3
 astroid==1.6.6            # via pylint, pylint-celery
 backports.functools-lru-cache==1.6.1  # via astroid, caniusepython3, isort, pylint
-backports.os==0.1.1       # via path.py
```
Comment (jmbowman) [inline requirements/dev.txt]: It looks like you ran `make upgrade` under Python 3.5, I think these are still needed under Python 2.

File: setup.cfg [hunk 21]
```diff
@@ -1,11 +1,19 @@
...
 known_edx =
 known_django = django
 known_djangoapp = model_utils
 known_first_party = edx_proctoring
+known_third_party=django,rest_framework,ddt,waffle,crum,httmock,edx_rest_framework_extensions,pytz,edx_when,eventtracking,opaque_keys,edx_rest_api_client,webpack_loader,jwt,freezegun,stevedore,Cryptodome
```
Comment (jmbowman) [inline setup.cfg]: Curious why you had to add this configuration, given that it's usually the default classification for packages outside the standard library.

File: tox.ini [hunk 22]
```diff
@@ -33,7 +40,8 @@ commands =
     rm -f docs/modules.rst
     make -C docs clean
     make -C docs html
-    python setup.py check --restructuredtext --strict
+    python setup.py sdist bdist_wheel
+    twine check dist/*
```
Comment (jmbowman) [inline tox.ini]: `twine` wasn't added to `requirements/doc.in`, and it looks like we don't run this code in Travis yet.  Probably should to avoid problems with package uploads on README verification.

File: edx_proctoring/models.py (hunk 16 above)
Comment (awais786) [reply edx_proctoring/models.py]:
```
/py27-django111-drf38/lib/python2.7/site-packages/six.py", line 935, in python_2_unicode_compatible
    klass.__name__)
ValueError: @python_2_unicode_compatible cannot be applied to ProctoredExamReviewPolicyHistory because it doesn't define __str__().
```

File: requirements/base.in (hunk 19 above)
Comment (awais786) [reply requirements/base.in]: I moved that in constraints.

File: requirements/dev.txt (hunk 20 above)
Comment (jmbowman) [reply requirements/dev.txt]: Still need to get back these 2 Python-2-only dependency pins.

Comment (awais786) [reply requirements/dev.txt]: I ran `make upgrade` in python2 but some how its removing automatically.
